p2p/transport/webrtcprivate/transport.go

```diff
@@ -0,0 +1,501 @@
+package libp2pwebrtcprivate
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p/core/connmgr"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	pionlogger "github.com/pion/logging"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	libp2pwebrtc "github.com/libp2p/go-libp2p/p2p/transport/webrtc"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtcprivate/pb"
+	"github.com/libp2p/go-msgio/pbio"
+	"github.com/pion/webrtc/v3"
+	"go.uber.org/zap/zapcore"
+
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+const (
+	name                = "webrtcprivate"
+	maxMsgSize          = 4096
+	connectTimeout      = time.Minute
+	SignalingProtocol   = "/webrtc-signaling"
+	disconnectedTimeout = 20 * time.Second
+	failedTimeout       = 30 * time.Second
+	keepaliveTimeout    = 15 * time.Second
+	maxAcceptQueueLen   = 10
+)
+
+var (
+	log        = logging.Logger("webrtcprivate")
+	WebRTCAddr = ma.StringCast("/webrtc")
+)
+
+type transport struct {
+	host                   host.Host
+	rcmgr                  network.ResourceManager
+	webrtcConfig           webrtc.Configuration
+	gater                  connmgr.ConnectionGater
+	maxInFlightConnections int
+
+	mu       sync.Mutex
+	listener *listener
+}
+
+var _ tpt.Transport = &transport{}
+
+func AddTransport(h host.Host, gater connmgr.ConnectionGater, stunServers []webrtc.ICEServer) (*transport, error) {
```

Comment:
Exported function returning unexported type.

Why is this method needed? Why don't we have such a method for other transports?

Reply:
We have this method for circuit-v2 transport. We need this method for the same reason for webrtc private transport. 

The peer address is `<relay-addr>/p2p-circuit/webrtc`. This transport will first make a relayed connection with the peer, which is why it needs access to the host. After this point it establishes the connection over the signaling stream.